packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -57,4 +67,81 @@ contract FederatedAttestations is
   function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
     return (1, 1, 0, 0);
   }
+
+  function _isRevoked(address signer, uint256 time) internal returns (bool) {
+    if (revokedSigners[signer] > 0 && revokedSigners[signer] >= time) {
+      return true;
+    }
+    return false;
+  }
+
+  function lookupAttestations(string identifier, address[] trustedIssuers)
+    public
+    view
+    returns (Attestation[])
+  {
+    Attestation[] memory attestations = new Attestation[];
+    for (uint256 i = 0; i < trustedIssuers.length; i++) {
+      address trustedIssuer = trustedIssuers[i];
+      for (uint256 j = 0; j < identifierToAddresses[identifier][trustedIssuer].length; j++) {
+        Attestation memory attestation = identifierToAddresses[identifier][trustedIssuer][j];
+        if (!_isRevoked(attestation.signer, attestation.issuedOn)) {
+          attestations.push(attestation);
+        }
+      }
+    }
+    return attestations;
+  }
+
+  function lookupIdentifiersbyAddress(address account, address[] trustedIssuers)
+    public
+    view
+    returns (bytes32[])
+  {
+    bytes32[] memory identifiers = new bytes32[];
+    for (uint256 i = 0; i < trustedIssuers.length; i++) {
+      address trustedIssuer = trustedIssuers[i];
+      for (uint256 j = 0; j < addressToIdentifiers[account][trustedIssuer].length; j++) {
+        bytes32 memory identifier = addressToIdentifiers[account][trustedIssuer][j];
+        Attestation memory attestation = identifierToAddresses[identifier][trustedIssuer];
+        if (!_isRevoked(attestation.signer, attestation.issuedOn)) {
+          identifiers.push(identifier);
+        }
+      }
+    }
+    return identifiers;
+  }
+
+  function registerAttestation(bytes32 identifier, address issuer, Attestation attestation) public {
+    require(
+      msg.sender == attestation.account || msg.sender == issuer || msg.sender == attestation.signer
+    );
+    for (uint256 i = 0; i < identifierToAddresses[identifier][issuer].length; i++) {
+      require(identifierToAddresses[identifier][issuer][i] != attestation.account);
+    }
+    identifierToAddresses[identifier][issuer].push(attestation);
+    addressToIdentifiers[attestation.account][issuer] = identifier;
+  }
+
+  function deleteAttestation(bytes32 identifier, address issuer, address account) public {
+    require(msg.sender == attestation.account || msg.sender == issuer);
+    for (uint256 i = 0; i < identifierToAddresses[identifier][issuer].length; i++) {
+      if (identifierToAddresses[identifier][issuer][i].account == account) {
+        identifierToAddresses[identifier][issuer][i] = identifierToAddresses[identifier][issuer][identifierToAddresses[identifier][issuer]
+          .length -
+          1];
+        identifierToAddresses[identifier][issuer].pop();
```

Comment:
This is meant to delete the attestation in the array and then move the last element in the array to that empty spot, to avoid having empty elements in the array. Not sure if this is needed and if the added gas costs from the complexity is worth it